### PR TITLE
build: Add "bundled" feature used in build script

### PR DESCRIPTION
**Pull request by boustrophedon**
Hi!

I'm building an easy-to-use wrapper around libseccomp and your library is really great!

I wanted to build it with musl and found it was a bit annoying because my linux distribution doesn't include a static version of libseccomp, so I added a `bundled` feature like [rusqlite's](https://github.com/rusqlite/rusqlite).

I went ahead and wrote the code but there are a couple of issues I'm not sure about. Let me know what you think and if you even think it's worth having this feature at all!

1. The code I added downloads the source from the github release.
  a. Would you instead prefer, like rusqlite, to include a copy of the libseccomp source code with the distribution?
2. Maintenance
  a. If we stick with downloading, the libseccomp version and checksum need to be manually updated when new versions of libseccomp are released.
  b. If we move to bundling, the libseccomp source would need to be manually updated.
3. I'm not 100% sure what the pkg-config in the libseccomp build.rs is doing or if it needs to be modified. It doesn't seem like it since I can build, e.g. for i686-unknown-linux-musl and run the tests without an issue.
4. Currently the build.rs code requires curl, tar, sha256sum and make to build libseccomp. It's possible to replace pretty much all of them with native rust dependencies, at the expense of increasing the build dependencies and build time.
  a. It might be reasonable to replace curl with git since it's more likely git will be installed given that we're building a Rust package.
5. Testing. I've tested locally by e.g. adding the i686-unknown-linux-musl target with rustup and running the tests, but haven't added it as a step in the github workflow.

Commit message:

```
This allows libseccomp to be cross-compiled, linked with musl, etc.

Requires curl, tar, sha256sum to build when the feature is enabled. Adds
the cc crate as a dependency, which is only built when the bundled
feature is enabled.
```

### CODE REVIEW

**Comment by boustrophedon** [comment]
Ah, it looks like the builds failed because building libseccomp requires gperf, which isn't installed in the container.

**Comment by boustrophedon** [comment]
Thank you for showing me the script!
> I hope this script will be helpful for you. Just to confirm, is your motivation for this PR to be able to use transparently the libseccomp crate without using the above script, even in environments that do not have the libseccomp library?

Yes, this is mostly correct. Consider the following use-case:
- You are building an application which depends on libseccomp-rs
- Your build environment doesn't contain a static libseccomp, or your target environment is on a different platform (e.g. arm)

In this case, you currently need to include in your build step something like `git pull https://github.com/libseccomp-rs/libseccomp-rs` and then run the build script, prior to running `cargo build --target aarch64-unknown-linux-musl` for your own application.

With the PR I've attached, instead you can just add `libseccomp = { version = "0.2", features = ["bundled"] }` to your Cargo.toml and Cargo will build libseccomp automatically when it builds the dependencies.

Do you think calling the existing script from build.rs would be a preferable option?

File: libseccomp-sys/build.rs
Comment (ManaSugi) [inline libseccomp-sys/build.rs:46]: Could you use `expect()` instead of `unwrap` for providing good error message?

**Review by ManaSugi — CHANGES_REQUESTED** [review]
Thank you for explaining in detail.
I understand your motivation and I think this feature will be useful for users!
@rusty-snake What do you think? 

> Do you think calling the existing script from build.rs would be a preferable option?

No, the existing script is basically for CI and developers, so I think the `build.rs` should not use the script.

I added some comments as the first review.

In addition, this feature will trigger license discussion because the libseccomp crate links the libseccomp library statically by default that is LGPL 2.0 if users turn on the feature. I'd like to stay out license discussion because I'm not a lawyer, but it would be better to add an explanation about it with how to use this feature in README to avoid confusion. Regarding this, I'll prepare for the explanation.

**Comment by boustrophedon** [comment]
Thank you for all the comments! I'll make all the necessary changes.

One thing I wasn't sure about was the "Signed-off-by" line. Should I add that myself or does it need to be done by a committer? Or both?

Regarding the licensing issue I also wasn't sure about that so thank you for taking care of it on your end.

File: libseccomp-sys/build.rs
Comment (boustrophedon) [reply libseccomp-sys/build.rs:130]: I'm okay to do this, but it seems also possible to do this by putting the values into a cargo config at `libseccomp-sys/.cargo/config.toml` and then reading the values from environment variables.

The benefit is we don't have to include extra dependencies to read from configuration files, but the main drawback is that the file is a bit hidden and easy to forget. This could be mitigated by including the details of upgrading the bundled libseccomp version in a "Maintenance" section in the readme, and making sure to leave a comment at the point of use about where the environment variable is set.

 (see [the cargo config documentation](https://doc.rust-lang.org/nightly/cargo/reference/config.html#env)] for more information)

Comment (rusty-snake) [reply libseccomp-sys/build.rs:130]: TBH a ini-like parser is written in a few lines like (not finish):

```rust
fn from_str(s: &str) -> HashMap<String, HashMap<String, String>> {
    let mut versions = HashMap::new();
    let mut cur_version;
    for line in s.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with("#") {
            continue;
        }
        
        if line.starts_with("[") {
            cur_version = versions
                .entry(line.trim_matches(&['[', ']'][..]).to_string())
                .or_default();
        } else {
            if let Some((k, v)) = line {
                cur_version.insert(k, v)
            }
        }
    }
    versions
}
```

IMHO we don't need more than a simple key=value format.

Comment (rusty-snake) [reply libseccomp-sys/build.rs:130]: BTW we can also manage this in a 

versions.rs

```rust
static LIBSECCOMP: SourceInfo = SourceInfo {
    description: "High level interface to Linux seccomp filter",
    url: "https://github.com/seccomp/libseccomp/releases/download",
    version: "2.5.3",
    sha256sum: "59065c8733364725e9721ba48c3a99bbc52af921daf48df4b1e012fbc7b10a76",
};
```

and `include!("versions.rs")` it.

**Comment by ManaSugi** [comment]
> One thing I wasn't sure about was the "Signed-off-by" line. Should I add that myself or does it need to be done by a committer? Or both?

Could you add "Signed-off-by" using `git commit --signoff` by yourself?

File: libseccomp-sys/Cargo.toml
Comment (ManaSugi) [reply libseccomp-sys/Cargo.toml:19]: >Because we're not using the cc crate to actually compile libseccomp, only to get cflags matching the cargo profile (i.e. --debug vs --release), I don't think using the parallel feature will make a difference here. I checked and it does not change the contents of the cflags.

Oh, I misunderstood, thank you for the explanation. 

 > Per the information in [the environment variables documentation for NUM_JOBS](https://doc.rust-lang.org/cargo/reference/environment-variables.html), I was able to set the MAKEFLAGS environment variable to the value of CARGO_MAKEFLAGS, which includes the -j option, and it did reduce the compile time when testing locally.
I'll include this change in the next revision.

Greate, it would be better to set the make env variable.

File: libseccomp-sys/build.rs
Comment (boustrophedon) [reply libseccomp-sys/build.rs:85]: While looking into the gperf issue, I now agree that it would be hard to build libseccomp using only cc because libseccomp uses gperf to generate syscall tables. However, because it only uses it at compile time, perhaps we should just require the user to install it via their package manager; the point of this build script is to bundle everything that's necessary at runtime, not at build time.

That is, I think it's fine to require the user also have gperf in the same way they need to have curl and tar and sha256sum installed in their build environment, but not necessarily in their target environment. Similarly, we're supporting the user in cross-compiling, but we're not downloading the compiler for them.

We could add a simple check to see if it's installed and have a nice error message instead of waiting for the make command to fail.

File: libseccomp-sys/build.rs
Comment (boustrophedon) [reply libseccomp-sys/build.rs:128]: Please see the above comment about gperf

File: libseccomp-sys/build.rs
Comment (boustrophedon) [reply libseccomp-sys/build.rs:130]: I think the versions.rs is a great solution - I'll implement that as well. I might not be able to push all changes until Tuesday or later this week because I have some job interviews this week (if you're hiring let me know!).

**Comment by rusty-snake** [comment]
If we need some reference

- https://crates.io/crates/autotools
- https://github.com/servo/libexpat/
- https://github.com/gyscos/zstd-rs/tree/master/zstd-safe/zstd-sys
- https://github.com/NoXF/libwebp-sys
- https://github.com/koutheir/sensors-sys
- https://github.com/koutheir/selinux-sys
- https://github.com/alexcrichton/openssl-src-rs
- https://github.com/sfackler/rust-openssl/tree/master/openssl-sys

File: libseccomp-sys/build.rs
Comment (boustrophedon) [reply libseccomp-sys/build.rs:85]: I tried uninstalling gperf and the error message seems good enough that we can probably leave it, actually.

```
  --- stderr
  configure: error: please install gperf
  Error: "Running `configure` failed while building libseccomp."
```

**Comment by boustrophedon** [comment]
I've pushed changes that should address all comments.

A few things remaining:
- Do we need to set the host and target flags in `configure` when cross compiling? Cross-compiling probably doesn't need to block the merge of this commit as it's useful on its own for making statically linked builds anyway.
- Build script modifications
  - Need to install gperf in the github build configuration so `cargo clippy --all-features` works.
  - Need to modify the github build configuration to test the bundled feature.
    - I've been checking that it's statically compiled by looking for libseccomp in the output of ldd.

File: CHANGELOG.md
Comment (rusty-snake) [inline CHANGELOG.md:37]:
```suggestion
- "bundled" feature to allow downloading and building libseccomp at build time.
```

File: libseccomp-sys/build.rs
Comment (rusty-snake) [inline libseccomp-sys/build.rs:11]: If we use a versions mod, we can

```suggestion
mod versions;
```

Comment (boustrophedon) [reply libseccomp-sys/build.rs:11]: I really thought I tried this and it didn't work inside a build script, but I just checked again and it does. Maybe I was tired and wrote "version.rs" instead of versions, or something like that. Will fix.

**Comment by boustrophedon** [comment]
Sorry for all the build failures. I tried adding an option to the install script to pass the gperf path to the action's $PATH variable but it didn't seem to work. [This was the documentation I looked at](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#example-8)

**Comment by boustrophedon** [comment]
Okay so it looks like there are a few ways I can pass the install script-built gperf path through to the following steps but they're all kind of a pain.

Instead I've just installed gperf via apt and now the build works. The downside is that we're building gperf even when we've already installed it via apt.

Should I add a flag to not build gperf when running the install script or do you think it's better to keep it simple?

**Review by ManaSugi — COMMENTED** [review]
Thank you for the update:)
I've done a quick review but will look at this more closely by the end of the week.

File: libseccomp-sys/versions.rs
Comment (rusty-snake) [reply libseccomp-sys/versions.rs:18]:
```rust
pub static LIBSECCOMP: SourceInfo = SourceInfo {
    description: "High level interface to Linux seccomp filter",
    url: "https://github.com/seccomp/libseccomp/releases/download/v%VERSION%/libseccomp-%VERSION%.tar.gz"    
    version: "2.5.3",
    sha256sum: "59065c8733364725e9721ba48c3a99bbc52af921daf48df4b1e012fbc7b10a76",
};

...

LIBSECCOMP.url.replace("%VERSION%", LIBSECCOMP.version)
```

**Comment by boustrophedon** [comment]
I've finally figured out what the issue is with exposing the script-built gperf to the following steps. I'm sorry it took so many builds and filled up your inboxes. Again, please don't feel rushed to review since it's not urgent at all.

When we execute the install script with sudo, the gperf directory is created with mktemp, which as the manpage states means that the directory is created u+rwx. Therefore the gperf output directory is not readable by non-root users.

So even after I passed the $GITHUB_PATH environment variable through sudo with -E and added the gperf bin directory to it, the following steps couldn't read from the directory into which gperf was built, because they aren't executed with sudo.

Since the script is intended to both be used on developer machines as well as in CI, I just made the temp dir readable by everyone.

If anyone is curious, here's how I ended up figuring it out, step by step: https://github.com/boustrophedon/github-actions-path-test/actions

File: libseccomp-sys/build.rs
Comment (ManaSugi) [inline libseccomp-sys/build.rs:24]: I think this bundled feature should be fired only when the target is musl, so it would be better to output error message if the target is not musl. What do you think?
@boustrophedon @rusty-snake

**Review by ManaSugi — COMMENTED** [review]
Thank you for the updates.
Could you replace the https://github.com/libseccomp-rs/libseccomp-rs/blob/main/libseccomp-sys/README.md#version-information with the link to `versions.rs`?
In addition, Could you add the description of this feature to README like https://github.com/rusqlite/rusqlite#optional-features?

File: .github/workflows/build-test.yaml
Comment (boustrophedon) [reply .github/workflows/build-test.yaml:89]: I just noticed I didn't actually use matrix.target when installing the toolchain, will include this fix on next push.

File: scripts/install_libseccomp.sh
Comment (boustrophedon) [reply scripts/install_libseccomp.sh:40]: I'll make the change to only execute build_and_install_gperf if it isn't installed, and use apt to install gperf in CI.

Thank you again for working with me on this feature and taking the time to review everything!

File: libseccomp-sys/build.rs
Comment (boustrophedon) [reply libseccomp-sys/build.rs:24]: It's entirely possible to use the bundled feature with glibc:

```
[hcs libseccomp]$ cargo test --features bundled -- --test-threads=1
[...]
[hcs libseccomp]$ ldd ../target/debug/deps/libseccomp-7f04e6b794ba92eb
	linux-vdso.so.1 (0x00007ffe2798c000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007faf6eeb9000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007faf6edd1000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007faf6ebc7000)
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007faf6f023000)
```

For example, if you are writing an application where you want to statically link libseccomp but dynamically link a proprietary dependency that requires glibc (and then use seccomp to make sure the dependency isn't doing anything bad!), it would be very hard to make everything work properly.

**Comment by boustrophedon** [comment]
I'd like to squash the two commits here into one if everything looks good, and then I guess we can merge?

File: libseccomp-sys/build.rs
Comment (ManaSugi) [inline libseccomp-sys/build.rs:26]: Cloud you leave a comment that static linking is done by default in the bundled feature?

File: README.md
Comment (ManaSugi) [inline README.md:119]: Cloud you fix this as follows? (We need to tell users that this feature **statically** links the libseccomp library by default)

### Building and statically linking the libseccomp library with the `bundled` feature

By enabling the bundled feature (see below), the libseccomp crate will automatically download,
build, and **statically** link the [libseccomp library ](https://github.com/seccomp/libseccomp) for you. 

```toml
[dependencies]
libseccomp = { version = "0.2.3", features = ["bundled"] }
```

The bundle feature downloads and builds the libseccomp sources according to the file [versions.rs](https://github.com/libseccomp-rs/libseccomp-rs/blob/main/libseccomp-sys/versions.rs)
by default if the `LIBSECCOMP_SRC_PATH` environment variable is not set. If you want to build
the libseccomp sources you already has on the system, set the `LIBSECCOMP_SRC_PATH`  to
specify the sources as follows.

```sh
$ export LIBSECCOMP_SRC_PATH="the path of the directory containing the libseccomp sources"
```

> **Note**: This feature links the libseccomp library statically by default.
> If you want to link the libseccomp library dynamically, you should not enable this feature.

File: scripts/install_libseccomp.sh
Comment (boustrophedon) [reply scripts/install_libseccomp.sh:35]: Thanks for catching these!

File: libseccomp-sys/build.rs
Comment (boustrophedon) [reply libseccomp-sys/build.rs:223]: Oh that's a good point - if you want to statically link with your own precompiled library, you can just pass LIB_PATH and LINK_TYPE without using the bundled feature. Will fix!

File: README.md
Comment (boustrophedon) [reply README.md]: I've updated the README files, let me know what you think!

File: scripts/install_libseccomp.sh
Comment (boustrophedon) [reply scripts/install_libseccomp.sh]: Wait actually isn't this still needed? At least one of the CI steps should not install gperf via apt so we can test this path, so I'll make that change.

Comment (boustrophedon) [reply scripts/install_libseccomp.sh]: https://github.com/libseccomp-rs/libseccomp-rs/runs/5382938044

yes, this line is still required when using the script to build and install gperf. In order to test this functionality, I'll remove the apt install gperf from the code coverage step since there's no matrix (1 run instead of 3) and the code coverage step takes longer anyway, so the increase in the step's runtime due to compiling gperf is smaller relative to the total time it takes for the step to complete.

Comment (boustrophedon) [reply scripts/install_libseccomp.sh]: ah, the coverage step uses --all-features and so we run into the same problem of needing to export the path to the next step so that the bundled build can find it.

I'll just have it run in the initial lint step then, unless you think we should add a separate step just to test the install script.

Comment (boustrophedon) [reply scripts/install_libseccomp.sh]: ... the clippy step also uses --all-features. I'll just add an extra step to test the build script.

Comment (boustrophedon) [reply scripts/install_libseccomp.sh]: Sorry for all the build failures. I ended up just adding a single job to test the install script by itself, and then the case where gperf is installed via apt is tested in the other jobs.

**Comment by boustrophedon** [comment]
Hi, are there still issues remaining that need to be fixed?

If the conflicting license is a legal issue and it seems like this won't be merged, I can try to move this code into my own project and use LIBSECCOMP_LIB_PATH.

**Comment by boustrophedon** [comment]
Okay, thanks for letting me know!

**Pull request closed without merging**
